test/e2e/storage/testsuites/volumelimits.go

```diff
@@ -0,0 +1,280 @@
+	// that says it's unschedulable because of volume limit.
+	// BEWARE: the test may create lot of volumes and it's really slow.
+	ginkgo.It("should support volume limits [Slow][Serial]", func() {
+		driverInfo := driver.GetDriverInfo()
+		if !driverInfo.Capabilities[CapVolumeLimits] {
+			ginkgo.Skip(fmt.Sprintf("driver %s does not support volume limits", driverInfo.Name))
+		}
+		var dDriver DynamicPVTestDriver
+		if dDriver = driver.(DynamicPVTestDriver); dDriver == nil {
+			ginkgo.Fail("Test driver does not provide dynamically created volumes")
+		}
+		claimSize := dDriver.GetClaimSize()
+
+		l.ns = f.Namespace
+		l.cs = f.ClientSet
+		l.config, l.testCleanup = driver.PrepareTest(f)
+		defer l.testCleanup()
+
+		ginkgo.By("Picking a random node")
+		var nodeName string
+		nodeList := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
+		if len(nodeList.Items) != 0 {
+			nodeName = nodeList.Items[0].Name
+		} else {
+			e2elog.Failf("Unable to find ready and schedulable Node")
+		}
+		e2elog.Logf("Selected node %s", nodeName)
+
+		ginkgo.By("Checking node limits")
+		// Wait in a loop, the driver might just have been installed and kubelet takes a while to publish everything.
+		var limit int
+		err := wait.PollImmediate(2*time.Second, csiNodeInfoTimeout, func() (bool, error) {
+			csiNode, err := l.cs.StorageV1beta1().CSINodes().Get(nodeName, metav1.GetOptions{})
+			if err != nil {
+				e2elog.Logf("%s", err)
+				return false, nil
+			}
+			var csiDriver *storagev1beta1.CSINodeDriver
+			for _, c := range csiNode.Spec.Drivers {
+				if c.Name == driverInfo.Name {
+					csiDriver = &c
+					break
+				}
+			}
+			if csiDriver == nil {
+				e2elog.Logf("CSINodeInfo does not have driver %s yet", driverInfo.Name)
+				return false, nil
+			}
+			if csiDriver.Allocatable == nil {
+				return false, fmt.Errorf("CSINodeInfo does not have Allocatable for driver %s", driverInfo.Name)
+			}
+			if csiDriver.Allocatable.Count == nil {
+				return false, fmt.Errorf("CSINodeInfo does not have Allocatable.Count for driver %s", driverInfo.Name)
+			}
+			limit = int(*csiDriver.Allocatable.Count)
+			return true, nil
+		})
+
+		e2elog.Logf("Node %s can handle %d volumes of driver %s", nodeName, limit, driverInfo.Name)
+
+		// Create a storage class and generate a PVC. Do not instantiate the PVC yet, keep it for the last pod.
+		l.resource = createGenericVolumeTestResource(driver, l.config, pattern)
+		defer l.resource.cleanupResource()
+
+		// Prepare cleanup
+		defer func() {
+			err = l.cs.CoreV1().Pods(l.ns.Name).Delete(l.runningPod.Name, nil)
+			framework.ExpectNoError(err, "failed cleaning up the running pod")
+			err = l.cs.CoreV1().Pods(l.ns.Name).Delete(l.unschedulablePod.Name, nil)
+			framework.ExpectNoError(err, "failed cleaning up the unschedulable pod")
+			for _, pvc := range l.pvcs {
+				err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Delete(pvc.Name, nil)
+				framework.ExpectNoError(err, "failed cleaning up pvc")
+			}
```

Comment:
If any of these is `err != nil`,  it won't try to delete the other objects?

Reply:
I reworked the cleanup completelly